diff -Naur react-native-bluetooth-state/android/build.gradle react-native-bluetooth-state-patched/android/build.gradle
--- react-native-bluetooth-state/android/build.gradle	2017-08-30 15:49:18.000000000 +0200
+++ react-native-bluetooth-state-patched/android/build.gradle	2018-12-06 23:23:34.000000000 +0100
@@ -1,13 +1,17 @@
 
 apply plugin: 'com.android.library'
 
+def safeExtGet(prop, fallback) {
+    rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
+}
+
 android {
-    compileSdkVersion 23
-    buildToolsVersion "23.0.1"
+    compileSdkVersion safeExtGet('compileSdkVersion', 27)
+    buildToolsVersion safeExtGet('buildToolsVersion', "27.0.3")
 
     defaultConfig {
-        minSdkVersion 16
-        targetSdkVersion 22
+        minSdkVersion safeExtGet('minSdkVersion', 16)
+        targetSdkVersion safeExtGet('targetSdkVersion', 26)
         versionCode 1
         versionName "1.0"
         ndk {
@@ -20,5 +24,5 @@
 }
 
 dependencies {
-    compile 'com.facebook.react:react-native:0.20.+'
+    implementation 'com.facebook.react:react-native:0.20.+'
 }
